core-primitives/stf-executor/src/executor.rs

```diff
@@ -0,0 +1,441 @@
+
+		let trusted_call = TrustedCallSigned::new(
+			TrustedCall::balance_shield(root, account, amount),
+			nonce,
+			Default::default(), //don't care about signature here
+		);
+
+		Stf::execute(&mut state, trusted_call, calls).map_err::<Error, _>(|e| e.into())?;
+
+		self.state_handler.write(state, state_lock, shard).map_err(|e| e.into())
+	}
+}
+
+impl<OCallApi, StateHandler> StfUpdateState for StfExecutor<OCallApi, StateHandler>
+where
+	OCallApi: EnclaveAttestationOCallApi + EnclaveOnChainOCallApi + GetStorageVerified,
+	StateHandler: HandleState,
+{
+	fn update_states<PB>(&self, header: &PB::Header) -> Result<()>
+	where
+		PB: BlockT<Hash = H256>,
+	{
+		debug!("Update STF storage upon block import!");
+		let storage_hashes = Stf::storage_hashes_to_update_on_block();
+
+		if storage_hashes.is_empty() {
+			return Ok(())
+		}
+
+		// global requests they are the same for every shard
+		let state_diff_update: StateTypeDiff = self
+			.ocall_api
+			.get_multiple_storages_verified(storage_hashes, header)
+			.map(into_map)?
+			.into();
+
+		// look for new shards an initialize them
+		if let Some(maybe_shards) = state_diff_update.get(&shards_key_hash()) {
+			match maybe_shards {
+				Some(shards) => {
+					let shards: Vec<ShardIdentifier> = Decode::decode(&mut shards.as_slice())?;
+
+					for shard_id in shards {
+						let (state_lock, mut state) =
+							self.state_handler.load_for_mutation(&shard_id)?;
+						trace!("Successfully loaded state, updating states ...");
+
+						// per shard (cid) requests
+						let per_shard_hashes = storage_hashes_to_update_per_shard(&shard_id);
+						let per_shard_update = self
+							.ocall_api
+							.get_multiple_storages_verified(per_shard_hashes, header)
+							.map(into_map)?;
+
+						Stf::update_storage(&mut state, &per_shard_update.into());
+						Stf::update_storage(&mut state, &state_diff_update);
+
+						// block number is purged from the substrate state so it can't be read like other storage values
+						// The number conversion is a bit unfortunate, but I wanted to prevent making the stf generic for now
+						Stf::update_layer_one_block_number(
+							&mut state,
+							(*header.number()).unique_saturated_into(),
+						);
+
+						self.state_handler.write(state, state_lock, &shard_id)?;
+					}
+				},
+				None => debug!("No shards are on the chain yet"),
+			};
+		};
+		Ok(())
+	}
+}
+
+impl<OCallApi, StateHandler> StfExecuteTimedCallsBatch for StfExecutor<OCallApi, StateHandler>
+where
+	OCallApi: EnclaveAttestationOCallApi + EnclaveOnChainOCallApi + GetStorageVerified,
+	StateHandler: HandleState,
+{
+	type Externalities = SgxExternalities;
+
+	fn execute_timed_calls_batch<PB, F>(
+		&self,
+		trusted_calls: &[TrustedCallSigned],
+		header: &PB::Header,
+		shard: &ShardIdentifier,
+		max_exec_duration: Duration,
+		prepare_state_function: F,
+	) -> Result<ExecutionResult>
+	where
+		PB: BlockT<Hash = H256>,
+		F: FnOnce(Self::Externalities) -> Self::Externalities,
+	{
```

Comment:
Execute a batch of trusted calls with a given time budget and return an `ExecutionResult` which includes the previous state hash, the execution status, and more information.